### ISSUE TITLE
Deadlocks with a blocking connection pool and retry

### ISSUE THREAD

**Issue by mattbillenstein** [shazow/urllib3#1167]
I started digging through this in requests via this issue:  https://github.com/kennethreitz/requests/issues/3993

Testcase here:  https://github.com/mattbillenstein/requests-retry-pool-hang

urllib3 seems to be leaking requests in this case and with a blocking connection pool, this causes a deadlock.

**Comment by Lukasa** [comment]
So the specific issue here is that urllib3 can leak connections during retry or redirect. For situations where we exceed our retry count, we manually release the connection because we aren't returning it, but we *do not* do that if we didn't exceed the count, even though in that situation we also aren't returning the connection. That seems obviously wrong.

This should be a nice fix, with a few tests required. Any contributor interested?

**Comment by mattbillenstein** [comment]
Here is what I've been testing with as a fix:  https://github.com/mattbillenstein/requests/commit/524b43154553489bf60216acb9030ce5b36f7215

This seems safe, since at these points, we'd never return that response anyway, so freeing anything the response relies on should be safe.

I will also say, it feels a bit weird to me that urlopen recurses in these cases -- and it is a pretty gnarly function --  but I have written code myself which does something similar.  Perhaps the retry/redirect logic should be in a function that wraps what urlopen is doing?  It's an idea anyway.

**Comment by Lukasa** [comment]
It's not *necessarily* safe: if the response body hasn't been consumed then that might cause the loss of a TCP connection we could keep. Some thought needs to go into that. 

As to the urlopen recursion, it's not really an issue: the retry count is never going to be large enough to cause a problem with the stack. However, the complexity of the code is a concern and there are some minor refactorings going on for v2.

**Comment by mattbillenstein** [comment]
Yeah, it had occurred to me that perhaps we were putting a dirty connection back to the pool -- maybe it's just safer to close the connection in case of a retry anyway -- a redirect, maybe that isn't desirable.

**Comment by mattbillenstein** [comment]
fyi, been running a version of this in prod that discards the connections before recursing -- at least for me, that seems to work pretty well -- ymmv

**Comment by Lukasa** [comment]
Yup, still totally happy to accept a patch to resolve this in the form of a PR. :smile:

**Comment by mattbillenstein** [comment]
Created https://github.com/shazow/urllib3/pull/1171

Not sure to craft a test for this...

**Comment by mattbillenstein** [comment]
Fixed in #1174
